include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+
+/**
+ * This callback informs the application that if data is added to the send
+ * buffer, some of it will be transmitted immediately without delay, as opposed
+ * to being queued for transmission at a later time. After a call to
+ * @p otTcpActiveSendByReference, an @p otTcpSendReady callback is guaranteed
+ * to be called, either immediately (if the connection is already ready) or
+ * sometime in the future (once the connection becomes ready for more data).
+ *
+ * This callback is not necessary for correct TCP operation. If more data is
+ * added to the send buffer than can be transmitted without delay, it will
+ * simply be queued for transmission at a later time. This callback should be
+ * used only in cases where some assurance is desired that data added to the
+ * send buffer will be sent soon (e.g., TCP won't wait for the recipient to
+ * ACK some other data first before sending this data out). For example, you
+ * may use this callback if you'd rather have your data be dropped than develop
+ * a backlog of data in your send buffer. But for most applications, where this
+ * isn't a concern, it's expected that one would not use this callback at all.
+ */
+typedef void (*otTcpSendReady)(void *aContext, otTcpActiveSocket *aSocket);
+
+/**
+ * This callback indicates the number of bytes available for consumption from
+ * the receive buffer. It is called whenever bytes are added to the receive
+ * buffer. If the end of the stream has been reached (i.e., if no more data
+ * will become available to read because the connection peer has closed their
+ * end of the connection for writing), then @p aEndOfStream is true. Finally,
+ * @p aBytesRemaining indicates how much capacity is left in the receive
+ * buffer to hold additional data that arrives.
+ */
+typedef void (*otTcpReceiveAvailable)(void *aContext, otTcpActiveSocket *aSocket, size_t aBytesAvailable, bool aEndOfStream, size_t aBytesRemaining);
+
+
+/**
+ * This callback indicates that the connection was broken and should no longer
+ * be used. It can occur if a connection establishment attempt (initiated by
+ * calling @p otTcpActiveConnect) fails, or any point thereafter (e.g., if the
+ * connection times out or an RST segment is received from the connection
+ * peer. Once this callback fires, all resources that the application provided
+ * for this connection (i.e., any `otLinkedBuffers` and memory they reference,
+ * but not the socket itself or space for the receive buffers) can be
+ * reclaimed.
+ */
+typedef void (*otTcpConnectionLost)(void *aContext, otTcpActiveSocket *aSocket, otError aErrorCode);
+
+/**
+ * This structure represents an active TCP socket.
+ *
+ * An active TCP socket acts an endpoint of TCP connection. It can be used to
+ * initiate TCP connections, and, once a TCP connection is established, send
+ * data to and receive data from the connection peer.
+ */
+typedef struct otTcpActiveSocket
```

Comment:
Unfortunately, I don't think this is possible. The reason is that the struct contains the callback function pointers, so the struct can only come after those `typedef`s. There is a bit of a circular dependency since the callback functions take arguments of type `otTcpActiveSocket *`, but I solved it via a forward declaration of `otTcpActiveSocket`.